usr/lib/common/mech_ec.c

```diff
@@ -1113,3 +1180,132 @@ CK_RV ecdh_pkcs_derive(STDLL_TokData_t *tokdata, SESSION *sess,
 
     return rc;
 }
+
+static int ec_nid_from_oid(CK_BYTE *oid, CK_ULONG oid_length)
+{
+    int i;
+
+    for (i = 0; i < NUMEC; i++) {
+        if (der_ec_supported[i].data_size == oid_length &&
+            memcmp(der_ec_supported[i].data, oid, oid_length) == 0)
+            return der_ec_supported[i].nid;
+    }
+
+    return -1;
+}
+
+/*
+ * Uncompress a compressed EC public key. EC public keys can be un-compressed,
+ * compressed, or hybrid. The fist byte of an EC public key determines if it
+ * is compressed or not:
+ * POINT_CONVERSION_COMPRESSED = 0x02
+ * POINT_CONVERSION_UNCOMPRESSED = 0x04
+ * POINT_CONVERSION_HYBRID = 0x06
+ * Bit 0x01 determines if it is odd or even
+ * The out_pubkey buffer size must be at least 1+2*privkey_len.
+ */
+CK_RV ec_uncompress_public_key(CK_BYTE *curve, CK_ULONG curve_len,
+                               CK_BYTE *pubkey, CK_ULONG pubkey_len,
+                               CK_ULONG privkey_len,
+                               CK_BYTE *out_pubkey, CK_ULONG *out_len)
+{
+    EC_GROUP *group = NULL;
+    EC_POINT *point = NULL;
+    CK_ULONG pad_len = 0;
+    BIGNUM *bn_x = NULL;
+    BIGNUM *bn_y = NULL;
+    BN_CTX *ctx = NULL;
+    CK_RV rc = CKR_OK;
```

Comment:
i generally prefer to initialize rc to some some default general error and set rc=CKR_OK just above the function's exit label (at that point its sure the function really was succesful)

Reply:
fixed